File: typescript/packages/core/src/commands/cli/builtin/gh/issue.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { FlagView } from '../../../spec/flag_view.ts'
import type { CommandFnResult } from '../../../config.ts'
import type { CLIInvocation } from '../../types.ts'
import {
  commentIssue,
  createIssue,
  editIssue,
  getIssue,
  listIssues,
  issueComments,
} from '../../../../core/github/issue.ts'
import {
  bodyValue,
  camel,
  csvValues,
  ghTransport,
  repoFor,
  repoNumber,
  textOut,
  textValue,
  typedOut,
} from './accessor.ts'

export const ISSUE_FIELDS = [
  'assignees',
  'author',
  'body',
  'closed',
  'createdAt',
  'labels',
  'number',
  'state',
  'title',
  'updatedAt',
  'url',
] as const

function issue(value: unknown): Record<string, unknown> {
  const row = camel(value)
  const result = row !== null && typeof row === 'object' ? (row as Record<string, unknown>) : {}
  result.closed = textValue(result.state).toLowerCase() === 'closed'
  return result
}

function names(value: unknown, key: 'login' | 'name'): string[] {
  if (!Array.isArray(value)) return []
  return value
    .filter((item): item is Record<string, unknown> => item !== null && typeof item === 'object')
    .map((item) => textValue(item[key] ?? item.name))
    .filter(Boolean)
}

function labels(row: Record<string, unknown>): string {
  return names(row.labels, 'name').join(', ')
}

function listText(rows: Record<string, unknown>[]): string {
  return rows
    .map(
      (row) =>
        `${textValue(row.number)}\t${textValue(row.state).toUpperCase()}\t${textValue(row.title)}\t${labels(row)}\n`,
    )
    .join('')
}

function viewText(row: Record<string, unknown>): string {
  const author = row.author as { login?: unknown } | undefined
  return `title:\t${textValue(row.title)}\nstate:\t${textValue(row.state).toUpperCase()}\nauthor:\t${textValue(author?.login)}\nlabels:\t${labels(row)}\n--\n${textValue(row.body)}\n`
}

function target(inv: CLIInvocation, fl: FlagView) {
  return repoNumber(inv, fl, inv.texts[0], 'issue', 'issues')
}

export async function listCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const params: Record<string, string> = { state: fl.asStr('state') ?? 'open' }
  const assignee = fl.asStr('assignee')
  const author = fl.asStr('author')
  const labelValues = csvValues(fl.asList('label'))
  if (assignee !== undefined) params.assignee = assignee
  if (author !== undefined) params.creator = author
  if (labelValues.length > 0) params.labels = labelValues.join(',')
  const values = await listIssues(
    ghTransport(inv.config),
    repoFor(inv, fl),
    params,
    fl.asInt('limit') ?? 30,
  )
  const rows = values.map(issue)
  return typedOut(rows, fl, listText(rows), ISSUE_FIELDS)
}

export async function viewCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const [ref, number] = target(inv, fl)
  const row = issue(await getIssue(ghTransport(inv.config), ref, number))
  const comments = await commentsFor(inv, fl, ref, number)
  if (comments !== null) row.comments = comments
  return typedOut(row, fl, fl.asBool('comments') ? commentsText(comments ?? []) : viewText(row), [
    ...ISSUE_FIELDS,
    'comments',
  ])
}

export async function createCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const title = fl.asStr('title')
  if (title === undefined || title === '')
    throw new Error('--title is required in noninteractive mode')
  const body: Record<string, unknown> = { title, body: (await bodyValue(inv, fl)) ?? '' }
  const labelValues = csvValues(fl.asList('label'))
  const assignees = csvValues(fl.asList('assignee'))
  if (labelValues.length > 0) body.labels = labelValues
  if (assignees.length > 0) body.assignees = assignees
  const created = issue(await createIssue(ghTransport(inv.config), repoFor(inv, fl), body))
  return textOut(`${textValue(created.url)}\n`)
}

export async function editCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const transport = ghTransport(inv.config)
  const [ref, number] = target(inv, fl)
  const body: Record<string, unknown> = {}
  const title = fl.asStr('title')
  const text = await bodyValue(inv, fl)
  if (title !== undefined) body.title = title
  if (text !== undefined) body.body = text
  const additions = csvValues(fl.asList('add_label'))
  const removals = new Set(csvValues(fl.asList('remove_label')))
  const addAssignees = csvValues(fl.asList('add_assignee'))
  const removeAssignees = new Set(csvValues(fl.asList('remove_assignee')))
  if (
    additions.length > 0 ||
    removals.size > 0 ||
    addAssignees.length > 0 ||
    removeAssignees.size > 0
  ) {
    const current = issue(await getIssue(transport, ref, number))
    if (additions.length > 0 || removals.size > 0) {
      body.labels = [
        ...new Set([
          ...names(current.labels, 'name').filter((name) => !removals.has(name)),
          ...additions,
        ]),
      ]
    }
    if (addAssignees.length > 0 || removeAssignees.size > 0) {
      body.assignees = [
        ...new Set([
          ...names(current.assignees, 'login').filter((name) => !removeAssignees.has(name)),
          ...addAssignees,
        ]),
      ]
    }
  }
  if (Object.keys(body).length === 0) throw new Error('no issue fields to edit')
  const edited = issue(await editIssue(transport, ref, number, body))
  return textOut(`${textValue(edited.url)}\n`)
}

async function state(inv: CLIInvocation, next: 'open' | 'closed'): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const [ref, number] = target(inv, fl)
  const edited = issue(await editIssue(ghTransport(inv.config), ref, number, { state: next }))
  const verb = next === 'closed' ? 'Closed' : 'Reopened'
  return textOut(
    `✓ ${verb} issue ${ref.owner}/${ref.repo}#${String(number)} (${textValue(edited.title)})\n`,
  )
}

export async function closeCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  return state(inv, 'closed')
}

export async function reopenCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  return state(inv, 'open')
}

export async function commentCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const body = await bodyValue(inv, fl, { required: true })
  const [ref, number] = target(inv, fl)
  const comment = issue(await commentIssue(ghTransport(inv.config), ref, number, body ?? ''))
  return textOut(`${textValue(comment.url)}\n`)
}

export async function commentsFor(
  inv: CLIInvocation,
  fl: FlagView,
  ref: { owner: string; repo: string },
  number: number,
): Promise<Record<string, unknown>[] | null> {
  if (!fl.asBool('comments') && !(fl.asStr('json') ?? '').split(',').includes('comments'))
    return null
  const rows = await issueComments(ghTransport(inv.config), ref, number)
  return rows.map((row) => ({
    ...row,
    author: row.author ?? { login: '' },
    minimizedReason: row.minimizedReason ?? '',
    reactionGroups: (row.reactionGroups as { users: { totalCount: number } }[]).filter(
      (group) => group.users.totalCount > 0,
    ),
  }))
}

export function commentsText(rows: Record<string, unknown>[]): string {
  return rows
    .map(
      (row) =>
        `author:\t${textValue((row.author as { login: string } | null)?.login)}\nassociation:\t${textValue(row.authorAssociation).toLowerCase()}\nedited:\t${String(row.includesCreatedEdit)}\nstatus:\t${row.isMinimized ? textValue(row.minimizedReason).toLowerCase() : 'none'}\n--\n${textValue(row.body)}\n--\n`,
    )
    .join('')
}

File: python/mirage/commands/cli/builtin/gh/issue.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any, cast

from mirage.commands.cli.builtin.gh.accessor import (body_value, camel,
                                                     csv_values, list_limit,
                                                     repo_for, repo_number,
                                                     text_out, typed_out)
from mirage.commands.cli.types import CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.core.github.config import GhConfig
from mirage.core.github.issue import (comment_issue, create_issue, edit_issue,
                                      get_issue, issue_comments, list_issues)
from mirage.core.github.repo import RepoRef
from mirage.io.types import ByteSource, IOResult
from mirage.types import JsonValue

ISSUE_FIELDS = ("assignees", "author", "body", "closed", "createdAt", "labels",
                "number", "state", "title", "updatedAt", "url")


def _issue(value: JsonValue) -> dict[str, Any]:
    row = camel(value)
    result = row if isinstance(row, dict) else {}
    result["closed"] = str(result.get("state", "")).lower() == "closed"
    return result


def _labels(row: dict[str, Any]) -> str:
    labels = row.get("labels")
    if not isinstance(labels, list):
        return ""
    return ", ".join(
        str(item.get("name", "")) for item in labels if isinstance(item, dict))


def _list_text(rows: list[dict[str, Any]]) -> str:
    return "".join(f'{row.get("number", "")}\t'
                   f'{str(row.get("state", "")).upper()}\t'
                   f'{row.get("title", "")}\t{_labels(row)}\n' for row in rows)


def _view_text(row: dict[str, Any]) -> str:
    author = row.get("author")
    login = author.get("login", "") if isinstance(author, dict) else ""
    return (f'title:\t{row.get("title", "")}\n'
            f'state:\t{str(row.get("state", "")).upper()}\n'
            f'author:\t{login}\nlabels:\t{_labels(row)}\n--\n'
            f'{row.get("body", "")}\n')


def _target(inv: CLIInvocation[GhConfig], fl: FlagView) -> tuple[RepoRef, int]:
    return repo_number(inv, fl, inv.texts[0] if inv.texts else None, "issue",
                       "issues")


async def list_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    params = {"state": fl.as_str("state") or "open"}
    if fl.as_str("assignee"):
        params["assignee"] = fl.as_str("assignee") or ""
    if fl.as_str("author"):
        params["creator"] = fl.as_str("author") or ""
    labels = csv_values(fl.as_list("label"))
    if labels:
        params["labels"] = ",".join(labels)
    values = await list_issues(inv.config, repo_for(inv, fl), params,
                               list_limit(fl, 30))
    rows = [_issue(value) for value in values]
    return await typed_out(rows, fl, _list_text(rows), ISSUE_FIELDS)


async def view_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    ref, number = _target(inv, fl)
    row = _issue(await get_issue(inv.config, ref, number))
    comments = await comments_for(inv, fl, ref, number)
    if comments is not None:
        row["comments"] = comments
    return await typed_out(
        row, fl,
        comments_text(comments or []) if fl.as_bool("comments") else
        _view_text(row), (*ISSUE_FIELDS, "comments"))


async def create_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    title = fl.as_str("title")
    if not title:
        raise ValueError("--title is required in noninteractive mode")
    body: dict[str, JsonValue] = {
        "title": title,
        "body": await body_value(inv, fl) or "",
    }
    labels = csv_values(fl.as_list("label"))
    assignees = csv_values(fl.as_list("assignee"))
    if labels:
        body["labels"] = cast(JsonValue, labels)
    if assignees:
        body["assignees"] = cast(JsonValue, assignees)
    created = _issue(await create_issue(inv.config, repo_for(inv, fl), body))
    return text_out(f'{created.get("url", "")}\n')


def _names(value: Any) -> list[str]:
    return [
        str(item.get("login", item.get("name", ""))) for item in value
        if isinstance(item, dict)
    ] if isinstance(value, list) else []


async def edit_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    ref, number = _target(inv, fl)
    body: dict[str, JsonValue] = {}
    if fl.as_str("title") is not None:
        body["title"] = fl.as_str("title") or ""
    body_text = await body_value(inv, fl)
    if body_text is not None:
        body["body"] = body_text
    additions = csv_values(fl.as_list("add_label"))
    removals = set(csv_values(fl.as_list("remove_label")))
    add_assignees = csv_values(fl.as_list("add_assignee"))
    remove_assignees = set(csv_values(fl.as_list("remove_assignee")))
    if additions or removals or add_assignees or remove_assignees:
        current = _issue(await get_issue(inv.config, ref, number))
        if additions or removals:
            names = _names(current.get("labels"))
            body["labels"] = list(
                dict.fromkeys([name
                               for name in names if name not in removals] +
                              additions))
        if add_assignees or remove_assignees:
            names = _names(current.get("assignees"))
            body["assignees"] = list(
                dict.fromkeys(
                    [name for name in names if name not in remove_assignees] +
                    add_assignees))
    if not body:
        raise ValueError("no issue fields to edit")
    edited = _issue(await edit_issue(inv.config, ref, number, body))
    return text_out(f'{edited.get("url", "")}\n')


async def _state(inv: CLIInvocation[GhConfig],
                 state: str) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    ref, number = _target(inv, fl)
    edited = _issue(await edit_issue(inv.config, ref, number,
                                     {"state": state}))
    verb = "Closed" if state == "closed" else "Reopened"
    return text_out(f'✓ {verb} issue {ref.owner}/{ref.repo}#{number} '
                    f'({edited.get("title", "")})\n')


async def close_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    return await _state(inv, "closed")


async def reopen_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    return await _state(inv, "open")


async def comment_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    body = await body_value(inv, fl, required=True)
    ref, number = _target(inv, fl)
    comment = _issue(await comment_issue(inv.config, ref, number, body or ""))
    return text_out(f'{comment.get("url", "")}\n')


async def comments_for(inv: CLIInvocation[GhConfig], fl: FlagView,
                       ref: RepoRef,
                       number: int) -> list[dict[str, Any]] | None:
    if not fl.as_bool("comments") and "comments" not in (fl.as_str("json")
                                                         or "").split(","):
        return None
    rows = await issue_comments(inv.config, ref, number)
    for row in rows:
        row["author"] = row.get("author") or {"login": ""}
        row["minimizedReason"] = row.get("minimizedReason") or ""
        row["reactionGroups"] = [
            group for group in row["reactionGroups"]
            if group["users"]["totalCount"] > 0
        ]
    return rows


def comments_text(rows: list[dict[str, Any]]) -> str:
    parts = []
    for row in rows:
        author = (row.get("author") or {}).get("login", "")
        status = row["minimizedReason"].lower(
        ) if row["isMinimized"] else "none"
        parts.append(f'author:\t{author}\n'
                     f'association:\t{row["authorAssociation"].lower()}\n'
                     f'edited:\t{str(row["includesCreatedEdit"]).lower()}\n'
                     f'status:\t{status}\n--\n{row["body"]}\n--\n')
    return "".join(parts)
